### ISSUE TITLE
Usage of read_dof_values_plain()

### ISSUE THREAD

**Issue by nfehn** [exadg/exadg#386]
In comments https://github.com/exadg/exadg/issues/383#issuecomment-1466891274 and https://github.com/exadg/exadg/issues/383#issuecomment-1467580096, the question came up whether we use `read_dof_values_plain()` correctly in ExaDG, in particular the question whether the current implementation will support hanging nodes or will do wrong calculations in case of hanging nodes.

**Comment by kronbichler** [comment]
I think the general motivation for `read_dof_values_plain` is to make it compatible with inhomogeneous constraints, but it is not an optimal solution. Let me summarize my state of knowledge with given capabilities and limitations. A few years ago, I wrote a longer text of options in the outlook of step-37: https://www.dealii.org/developer/doxygen/deal.II/step_37.html#InhomogeneousDirichletboundaryconditions

As said in the other issue, `read_dof_values_plain` does not resolve any constraints, so also hanging node constraints or periodicity constraints are not resolved. This makes that concept fragile when used with adaptive mesh refinement. One can solve this by having an `AffineConstraints` object with up-to-date constraints, from which we can call `AffineConstraints::distribute()` on a global vector. This is a bit slow, and suffers from the fact that we do not want to update the `AffineConstraints` object just because the inhomogeneity changes (but not hanging nodes, see below). The options proposed in the step-37 discussion there also discuss the option of setting up `MatrixFree` with the same `DoFHandler` but different constraints objects, one with all constraints (including boundary conditions) and another one with only the "hard" constraints (hanging nodes, periodicity). However, also that setup is not super nice, because we need to distinguish them with a different `dof_index` in terms of the `MatrixFree` indices, which would lead to some changes in our data structures in ExaDG.

This ambiguity shows a general limitation in the current realization of `MatrixFree` in deal.II, in the sense that there is no mechanism to handle inhomogeneities out-of-the-box. The reason is that inhomogeneities are not so simple, for the following two reasons:
- Constraints need to be resolved before they can be used. (This happens inside `AffineConstraints::close()`, so simply providing a new set of inhomogeneous values from a call to `VectorTools::interpolate_boundary_values` to an already finalized `AffineConstraints` does not yield a compatible solution.) Think of the case of a hanging node constraint of the form `x_2 = 0.5 x_0 + 0.5 x_1` for an unknown `x_2` in terms of two other dofs (linear shape functions, 2D) and a Dirichlet boundary condition resulting in the inhomogeneous constraint `x_0 = 17` because `x_0` sits on a Dirichlet boundary. The final result in affine constraints will have the form `x_2 = 8.5 + 0.5 x_1`.
- The first issue could be fixed by providing a newly constructed  `AffineConstraints` object every time a boundary condition changes. However, both the construction of a complete `AffineConstraints` object as well as a hypothetical `MatrixFree::reinit()` (remembering the inhomogeneous constraints and being able to impose them correctly in a hypothetitcal `FEEvaluation::read_dof_values_inhomogeneous` function) are not so cheap that we would want to call them in every time step. Also, it feels wrong to do this when the bulk of the work is done with homogeneous constraints that do not even change.

The option of having two `AffineConstraints` object on the same `DoFHandler` was born out of this conflict: If we have two `AffineConstraints` objects where the second one does not include those DoF indices subject to inhomongeneous constraints, we can use one or the other for the homogeneous or inhomogeneous operator. The inhomogeneous operator needs to be combined with a call to `VectorTools::interpolate_boundary_values`, using the resulting std::map to set the entries in the global solution vector, and use `FEEvaluation::read_dof_values/gather_evaluate` from the `DoFHandler/AffineConstraints` pair with no constraints on the inhomogeneous boundaries to get the values on a cell. Conversely, we would use `FEEvaluation::read_dof_values/gather_evaluate` from the `DoFHandler/AffineConstraints` pair related to the larger set of constraints (including `DoFTools::make_zero_boundary_constraints`) for the homogeneous operator as we do today.

It would be great to have a simpler solution, but I have not found it yet. One could think about embedding better information information into the `FEEvaluation::read_dof_values_plain()` (better called `read_dof_values_without_inhomogeneous_constraints()`), but that means we would need to switch the underlying `dof_indices_plain` field with this choice. Other opinions?

**Comment by nfehn** [comment]
@kronbichler many thanks for your detailed explanations!

I might have another option (but it might also be just some naive thoughts ...). My goal is to avoid two `AffineConstraints` objects and to make the implementation easier / more intuitive than it is currently. I have the concern that two separate objects are too complicated in use (given that it seems as if we cannot avoid some hand-written code for explicit imposition of inhomogeneous boundary data in user code anyway).

I think the main problem is the circumstance that `dealii::AffineConstraints` combines two different things:
- constraints between dofs (hanging nodes, periodic boundaries)
- homogeneous and inhomogeneous boundary data

I consider this a design problem on the deal.II side. It might be that also `dealii::MatrixFree` kind of suffers from this early design choice, I don't know.

With the current realization in ExaDG, we have two problems:
- we use `read_dof_values_plain()` for nonlinear residual evaluation with non-trivial boundary conditions, which will not work for hanging nodes/periodic boundaries correctly. I did not know this, and based on what you have written, I am tempted to argue that `read_dof_values_plain()` should not be used in ExaDG.
- it does not seem easy to use `MatrixFree` with both homogeneous and inhomogeneous boundary data due to design decisions in deal.II. This seems, however, a standard use case in application projects like ExaDG. I fear not every ExaDG developer is able to understand some of the implicit logic behind the constraints in deal.II (e.g. during the last days, I have seen different places in the deal.II docu mentioning which constraints have to be imposed first, ...). To some extent, this also means that other researchers might hestitate to realize application projects (PDE models more complex than what we have currently in ExaDG) with ExaDG for this reason (and for other reasons as well ...).

Let me propose the following design as a potential procedure in ExaDG:
- use `dealii::AffineConstraints` only for what it can be used straight-forwardly with `dealii::MatrixFree`, i.e. hanging nodes / periodicity constraints, but not for homogeneous/inhomogeneous boundary data
- impose homogeneous / inhomogeneous boundary data explicitly in ExaDG, i.e. do not incorporate such boundary conditions in `dealiiAffine::Constraints` (and, as a consequence, also not in `dealii::MatrixFree/FEEvaluation`).
- then, using `read_dof_values()` always, we impose hanging node / periodicity constraints (which is currently not the case for nonlinear residual evaluation) and we read the homogeneous / inhomogeneous boundary data from the dof-vector when using `FEEvaluation::gather_evaluate()`, because `AffineConstraints/MatrixFree` is not aware of these boundary "constraints". Of course, we have to manipulate the the dof-vectors appriopriately prior to evaluation. However, I think one can provide utility functions for this and/or realize this in base classes such as `OperatorBase`, such that a concrete operator only has to implement the part where the boundary conditions are evaluated.
- it looks like we have most of the ingredients already realized in ExaDG: We have the `OperatorBase::apply()` function, see https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/operators/operator_base.cpp#L260 for the evaluation of the homogeneous part of a linear PDE operator which handles homogeneous boundary data explicitly (somewhat superfluous in the current code version because of the "redundant" AffineConstraints including homogeneous boundary data currently),  the functions `set_constrained_values_to_zero()`, see https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/operators/operator_base.cpp#L237 and  `set_constrained_values()`, see https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/operators/operator_base.cpp#L375 used to evaluate PDE operators with inhomogeneous boundary data (rhs-vector for linear problems, nonlinear residual evaluation for nonlinear problems). The function in `Structure::ElasticityOperatorBase::set_constrained_values()`, see https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/structure/spatial_discretization/operators/elasticity_operator_base.cpp#L108 shows a concrete implementation, using `dealii::VectorTools::interpolate_boundary_values()` with the `std::map<global_dof_index, ...>` argument.

In summary, it seems as we already do what you described above as the "second `AffineConstraints` object" (without inhomogeneous boundary data in order to not always reinit `MatrixFree`, i.e. for performance reasons, at the price of handling the inhomogeneous boundary constraints somewhat explicitly in application code). But then I wonder whether we need the "first `AffineConstraints` object" (with homogeneous boundary data) at all. Note that, at least from my current understanding, we already have the infrastructure in ExaDG to impose homogeneous boundary data explicitly, see
https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/operators/operator_base.cpp#L324-L331

**Comment by kronbichler** [comment]
> I consider this [combining hanging node constraints with Dirichlet data] a design problem on the deal.II side. It might be that also dealii::MatrixFree kind of suffers from this early design choice, I don't know.

I would not say that this is a design problem in deal.II: deal.II is a library providing a set of options to implement algorithms, and it is up to the user to use it in one or another way. So if we call it a problem, it is rather one with how I suggest to use the features, as the decision in ExaDG is based on my suggestions in the past.

The manual way to impose the conditions (a component #384 would remove) as you suggest it might look a solution (with a small performance penalty, but performance should not be our concern for now), but in fact it does not according to my understanding: If you create a matrix, e.g., for a coarse AMG solver, you want to impose both Dirichlet boundary conditions and hanging node constraints, and hence we would need to provide additional mechanisms to impose those conditions uniformly. But not even in the matrix-free context it has not turned out to be the silver bullet: Karl Ljungkvist and I have used this algorithm for GPU programming, for example. Splitting up the constraints into two parts makes the necessary loops for the homogeneous operator more complicated, even though hanging nodes/periodicity or Dirichlet boundary conditions really are the same for the homogeneous operator: They do not carry information for the linear system and should not couple to other unknowns in the system. So With the suggested code, the `constrained_indices` are needed in two flavors: The unknowns that need to be cleared before entering the cell loop and the ones to get the assignment `src.local_element(constrained_indices[i]) = dst.local_element(constrained_indices[i])` are different. The latter needs to be called on **both** the constrained indices of the first kind (hanging nodes, periodicity) and the second kind (Dirichlet), but the former should not. Conversely, setting them to zero only needs the code on the `constrained_indices` originating from Dirichlet, the rest is automatically zero. So while we shift away the problem from the `FEEvaluation/CellIntegrator` classes, we move it to the context of `cell_loop`.

From past experience and having tried it in more than one project, I am thus not yet convinced.

Let me add a historical note: When I started serious programming with deal.II in 2008, Dirichlet boundary conditions were not part of the `AffineConstraints` object (called `ConstraintMatrix` back then), but done in two steps. It was actually me who added the feature of inhomogeneous constraints (around 2009?), which enabled the combination of Dirichlet boundary conditions and hanging nodes for the linear system in one go. This did not solve all problems, but simplified the situation considerably for matrix-based simulations because keeping track of Dirichlet boundary conditions and when to apply them to get a consistent state is not easy. Maybe this is still reflected in part of the documentation that you found confusing. For `MatrixFree`, this concept was taken over, even though another solution might have been better.  `read_dof_values_plain` was introduced for the not-so-well-supported case of Dirichlet boundary conditions. One of the codes using it was https://github.com/kronbichler/adaflo/blob/926aa96b76f5a9af63890c891808d9e7310455f2/source/navier_stokes_matrix.cc#L662-L665, where we combined it with manually imposing hanging node constraints, https://github.com/kronbichler/adaflo/blob/926aa96b76f5a9af63890c891808d9e7310455f2/source/navier_stokes.cc#L1233-L1245

> This seems, however, a standard use case in application projects like ExaDG. I fear not every ExaDG developer is able to understand some of the implicit logic behind the constraints in deal.II (e.g. during the last days, I have seen different places in the deal.II docu mentioning which constraints have to be imposed first, ...). To some extent, this also means that other researchers might hestitate to realize application projects (PDE models more complex than what we have currently in ExaDG) with ExaDG for this reason (and for other reasons as well ...).

If we find a way to make the case with deal.II/`MatrixFree` uniformly better, we should definitely do that! However, as I explained above, I am not sure if we have this solution yet. Regarding other researchers working with ExaDG: If we provide a good solution with appropriate comments/logic behind, users will be able to understand the choices and imitate the existing modules, even if it requires some work. But part of the solution is also to not see deal.II as something static: We can change things, and if we have good ideas, we should do that.

So, let me end this long message with the question what we would actually like to have as an interface, and then see if that interface is possible. From my point of view, the ideal interface would be
- at setup time, I give an `AffineConstraints` object with the hard constraints (periodicity, hanging nodes, also Dirichlet boundary conditions that will always be zero and can never get inhomogeneous) and additionally some DoFs that will get inhomogeneous constraints, say an `std::map` (this is done for each of the `DoFHandler` objects)
- before calling a `MatrixFree::loop/cell_loop`, I need to be able to pass in an `std::map` (or some more efficient data structure) to have it internally applied
- we introduce a new function `FEEvaluation::read_dof_values_with_inhomogeneity`
- all `integrate` functions and the `read_dof_values` call use the homogeneous constraints.
What are other opinions?

**Comment by nfehn** [comment]
Thanks for your detailed reply @kronbichler.

>  So With the suggested code, the constrained_indices are needed in two flavors: The unknowns that need to be cleared before entering the cell loop and the ones to get the assignment src.local_element(constrained_indices[i]) = dst.local_element(constrained_indices[i]) are different. The latter needs to be called on both the constrained indices of the first kind (hanging nodes, periodicity) and the second kind (Dirichlet), but the former should not.

From your explanations it became clear to me that I did not understand this correctly. I thought `constrained_indices` refer to Dirichlet constraints only. This makes my suggestion probably a sub-optimal solution.

Your suggestion regarding new capabilities in deal.II sounds good. I assume that `FEEvaluation::read_dof_values_with_inhomogeneity()` will take into account hanging nodes / periodicity as well. From my (limited) experience with constraints in ExaDG, it seems as if the `std::map` for the inhomogeneous boundary data is what we need in addition to the current functionality. Are there reasons that you do not prefer the "status-quo data structure" used in deal.II, `std::map<global_dof_index, double>` e.g. in `dealii::VectorTools::interpolate_boundary_values()`? Some aspects are not entirely clear to me right now.

**Nonlinear problem:** If I understand you correctly, functions like `set_constrained_values()` are then no longer needed for nonlinear residual evaluation prior to calling `matrix_free->loop()`, because this information is somehow passed to `MatrixFree` and correctly evaluated by the new function for reading dof values. However, how would the following code done at the end of nonlinear residual evaluation in structural mechanics look like in the future

https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/structure/spatial_discretization/operator.cpp#L760-L764

Should it set zeros for all types of constraints (hanging nodes, periodicity, dirichlet)? Would this remain a utility function in application projects such as `ExaDG`?

**Linear problems with inhomogeneous boundary data:** Forming the right-hand side vector and solving the linear system of equations

https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/structure/spatial_discretization/operator.cpp#L846-L872

Again, I am not sure whether `set_constrained_values()` should also manipulate hanging-node / periodicity constraints in the future. We currently also have to manipulate the dof-vector after a `vmult()` call, see PR #384.

**Summary:** The mis-understanding I had above seems to be reflected also in the function names `OperatorBase::set_constrained_values_to_zero()` (addressing all `constrained_indices`) and `OperatorBase::set_constrained_values()` (imposing, at least currently, only inhomogeneous boundary data). Are the function names imprecise currently, or would we extend the second function to hanging nodes / periodicity in the future in order to deal with such problems that are currently not fully supported in ExaDG especially for CG discretizations?

The function `FEEvaluation::read_dof_values_with_inhomogeneity()` is something that I want to call a "pre-operation" here. Are similar "post-operations" (done at the end of matrix-vector product) needed as well as general utility in `deal.II/MatrixFree` in order to make writing application code easier? Or do we need something like `AffineConstraints::distribute()` for the new data structure for inhomogeneous boundary data (potentially also addressing hanging nodes / periodicity constraints, see my questions raised above), which then needs to be called in user code as we currently do in ExaDG?
